tools/clang/lib/SPIRV/EmitSPIRVAction.cpp

```diff
@@ -23,25 +23,90 @@
 
 namespace clang {
 namespace {
-class SPIRVEmitter : public ASTConsumer,
-                     public RecursiveASTVisitor<SPIRVEmitter> {
+
+class SPIRVEmitter : public ASTConsumer {
 public:
   explicit SPIRVEmitter(raw_ostream *Out)
-      : OutStream(*Out), TheContext(), Builder(&TheContext) {}
+      : OutStream(*Out), TheContext(), TheBuilder(&TheContext) {}
 
   void HandleTranslationUnit(ASTContext &Context) override {
-    Builder.beginModule();
-    Builder.endModule();
-    std::vector<uint32_t> M = Builder.takeModule();
+    TheBuilder.requireCapability(spv::Capability::Shader);
+
+    // Addressing and memory model are required in a valid SPIR-V module.
+    TheBuilder.setAddressingModel(spv::AddressingModel::Logical);
+    TheBuilder.setMemoryModel(spv::MemoryModel::GLSL450);
+
+    for (auto *decl : Context.getTranslationUnitDecl()->decls()) {
+      DoDecl(decl);
+    }
+
+    std::vector<uint32_t> M = TheBuilder.takeModule();
     OutStream.write(reinterpret_cast<const char *>(M.data()), M.size() * 4);
   }
 
+  void DoDecl(Decl *decl) {
+    if (auto *funcDecl = dyn_cast<FunctionDecl>(decl)) {
```

Comment:
We cannot do it right now since that will result in an assertion when translating `void main() {}`. (We need a better way like emitting diagnostics for unimplemented features.)